### PR TITLE
add order id to voucher and confirmation route

### PR DESCRIPTION
**Pull request by VincentGossey**

### CODE REVIEW

File: src/Helper/VoucherHelper.php [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    {
+        return $this->apiVoucherFactory->createAndValidateVoucherRequest($data);
+    }
+
+    /**
+     * @param ApiVoucherModel $voucherModel
+     * @return string|null
+     * @throws \Doctrine\ORM\NonUniqueResultException
+     */
+    public function availableStock(ApiVoucherModel $voucherModel): ?string
```
Comment (AymericJoubert) [inline src/Helper/VoucherHelper.php]: getAvailableStock

File: src/Service/StockAlertNotification.php [hunk 2]
```diff
@@ -26,6 +26,7 @@ public function __construct(ProductManager $productManager)
      */
     public function getStockAlertCount(): ?int
     {
-        return count($this->productManager->getProducts([ProductRepository::STOCK_ALERT_LEVEL_PARAM_NAME => true]));
+        $products = $this->productManager->getProducts([ProductRepository::STOCK_ALERT_LEVEL_PARAM_NAME => true]);
+        return $products != null ? count($products) : null;
```
Comment (AymericJoubert) [inline src/Service/StockAlertNotification.php]: ligne vide avant le return

File: src/Controller/Api/VoucherController.php [hunk 3]
```diff
@@ -136,52 +128,83 @@ public function __construct(ApiManager $apiManager)
...
-        $apiData =  $request->query->getIterator()->getArrayCopy();
-
-        if (count(array_diff_key($apiData, array_flip(self::API_PARAMS))) || array_search(null, $apiData)) {
-            return new JsonResponse(['error_code' => 4, 'message' => 'Bad Request'], JsonResponse::HTTP_BAD_REQUEST);
-        };
+    public function getVouchers(Request $request, EventDispatcherInterface $eventDispatcher): JsonResponse
+    {
+        // Throw an ApiRequestException and return json with corresponding code if an error occurs
+        $apiData = $this->voucherHelper->createVoucherRequest($request->query->getIterator()->getArrayCopy());
+        $this->voucherHelper->isStockAvailable($apiData);
```
[the remark this answers is not in the document]
Comment (VincentGossey) [reply src/Controller/Api/VoucherController.php]: Done

File: src/Controller/Api/VoucherController.php [hunk 4]
```diff
@@ -136,52 +128,83 @@ public function __construct(ApiManager $apiManager)
...
-        $availableStock = $this->apiManager->availableStock($merchantId, $voucherAmount);
-        if ($availableStock < $voucherCount) {
-            return new JsonResponse(
-                ['error_code' => 1, 'message' => 'Insufficient or empty stock', 'available stock' => $availableStock],
-                JsonResponse::HTTP_NOT_FOUND
-            );
-        }
+        $newAvailableStock = $this->voucherHelper->availableStock($apiData);
+        $vouchers = $this->setData($result, $newAvailableStock, $apiData->getOrderId());
+        return new JsonResponse($vouchers, JsonResponse::HTTP_OK);
```
[the remark this answers is not in the document]
Comment (VincentGossey) [reply src/Controller/Api/VoucherController.php]: ok

File: src/Controller/Api/VoucherController.php [hunk 5]
```diff
@@ -136,52 +128,83 @@ public function __construct(ApiManager $apiManager)
...
+        $eventDispatcher->dispatch(self::ALERT_STOCK_EVENT, new StockAlertEvent($result));
 
-        $availableStock = $this->apiManager->availableStock($merchantId, $voucherAmount);
-        if ($availableStock < $voucherCount) {
-            return new JsonResponse(
-                ['error_code' => 1, 'message' => 'Insufficient or empty stock', 'available stock' => $availableStock],
-                JsonResponse::HTTP_NOT_FOUND
-            );
-        }
+        $newAvailableStock = $this->voucherHelper->availableStock($apiData);
```
[the remark this answers is not in the document]
Comment (VincentGossey) [reply src/Controller/Api/VoucherController.php]: ok

File: src/EventListener/ApiLoggerListener.php [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+            );
+        }
+    }
+
+    private function getFirewallContext(): array
+    {
+        return [
+            'security.firewall.map.context.api',
+            'security.firewall.map.context.token',
+            ];
```
[the remark this answers is not in the document]
Comment (VincentGossey) [reply src/EventListener/ApiLoggerListener.php]: ok

File: src/Helper/VoucherHelper.php (hunk 1 above)
Comment (VincentGossey) [reply src/Helper/VoucherHelper.php]: ok

File: src/Service/StockAlertNotification.php (hunk 2 above)
Comment (VincentGossey) [reply src/Service/StockAlertNotification.php]: ok

**Pull request merged**